### ISSUE TITLE
ActivityWorker abstraction

### ISSUE THREAD

**Issue by sclasen** [sclasen/swfsm#10]
Come up with a good way/similar to FSM to write activity workers.

* short lived activities: no heartbeating required
* long lived: heartbeating/cooperative cancelation of ongoing work with the poller shutdown manager and the heartbeater.

**Comment by sclasen** [comment]
This should facilitate testing as well. its super verbose to test without using the SWF service, we dont generally unit test db apps without the DB, so we should be able to easily swap in a test activity worker that doesnt do any actual work but responds to activities (by failing often). Each activity should fail and timeout a few times before completing successfully, based on a `func(input inteface{}) (output interface{})`. 

So to test FSMs, we actually spin them up and start a workflow, but we start the test activity workers instead of the actual activity workers, and they fail a bunch and eventually complete.

We are then saved from the terrible task of writing all the testing boilerplate. We just have to simulate all the appropriate signals, which we can even send from the TestActivityWorkers at (in)appropriate times to test problematic interleavings.

```
type ActivityIO struct {
      Activity string
      Input  interface{}
      Output interace{}
}

type  ActivityWorker struct {
     HandledActivities []ActivityIO
}

type TestActivityWorker struct {
     ActivityWorker
     CompletionResults map[string]func(input inteface{}) (output interface{})
     FailureResults map[string]func(input inteface{}) (output interface{}) //if a failure is not defined for an activity, default to something
}

```

We would also then provide TypedFuncs to do the correct type coercion / serialization / deserialization of inputs and outputs to the activities.

**Comment by sclasen** [comment]
See #23

**Comment by sclasen** [comment]
closed by #23
